File: lib/builtins/flutter/material/inkWell.dart
import 'package:flua/luastate.dart';
import 'package:flua/vm/closure.dart';
import 'package:flua/vm/context.dart';
import 'package:flua/builtins/flutter/syntheticBox.dart';
import 'package:flua/vm/table.dart';
import 'package:flutter/material.dart';

loadInkWell({@required LuaState luaState, @required HydroTable table}) {
  table["inkWell"] = makeLuaDartFunc(func: (List<dynamic> args) {
    return [
      InkWell(
        key: maybeUnwrapAndBuildArgument<Key>(args[0]["key"],
            parentState: luaState),
        child: maybeUnwrapAndBuildArgument<Widget>(args[0]["child"],
            parentState: luaState),
        onTap: () {
          Closure closure = args[0]["onTap"];
          closure.dispatch([null], parentState: luaState);
        },
        focusColor: maybeUnwrapAndBuildArgument<Color>(args[0]["focusColor"],
            parentState: luaState),
        hoverColor: maybeUnwrapAndBuildArgument<Color>(args[0]["hoverColor"],
            parentState: luaState),
        highlightColor: maybeUnwrapAndBuildArgument<Color>(
            args[0]["highlightColor"],
            parentState: luaState),
        splashColor: maybeUnwrapAndBuildArgument<Color>(args[0]["splashColor"],
            parentState: luaState),
        radius: args[0]["radius"],
        borderRadius: maybeUnwrapAndBuildArgument<BorderRadius>(
            args[0]["borderRadius"],
            parentState: luaState),
        enableFeedback: args[0]["enableFeedback"],
        canRequestFocus: args[0]["canRequestFocus"],
        autofocus: args[0]["autoFocus"],
      )
    ];
  });
}

File: runtime/flutter/material/inkWell.ts
import { Type } from "../../dart/core/type";
import { Color } from "../../dart/ui/color";
import { RuntimeBaseClass } from "../../runtimeBaseClass";
import { Key } from "./../foundation/key";
import { BorderRadius } from "../painting/borderRadius";
import { Widget } from "../widget";
import { StatelessWidget } from "../widgets/statelessWidget";

export interface InkWellProps {
    key?: Key | undefined;
    child: Widget;
    onTap: () => void;
    focusColor?: Color | undefined;
    hoverColor?: Color | undefined;
    highlightColor?: Color | undefined;
    splashColor?: Color | undefined;
    radius?: number | undefined;
    borderRadius?: BorderRadius | undefined;
    enableFeedback?: boolean | undefined;
    excludeFromSemantics?: boolean | undefined;
    canRequestFocus?: boolean | undefined;
    autoFocus?: boolean | undefined;
}

declare const flutter: {
    material: {
        inkWell: (this: void, props: InkWellProps) => InkWell;
    };
};

export class InkWell extends StatelessWidget implements RuntimeBaseClass {
    public readonly internalRuntimeType = new Type(InkWell);
    public props: InkWellProps;
    public constructor(props: InkWellProps) {
        super();
        this.props = props;

        if (this.props.enableFeedback === undefined) {
            this.props.enableFeedback = true;
        }

        if (this.props.excludeFromSemantics === undefined) {
            this.props.excludeFromSemantics = false;
        }

        if (this.props.canRequestFocus === undefined) {
            this.props.canRequestFocus = true;
        }

        if (this.props.autoFocus === undefined) {
            this.props.autoFocus = false;
        }
    }

    public build() {
        return flutter.material.inkWell(this.props);
    }
}
